### DISCUSSION TITLE
[Chapter 05] ControllerAdvice 예외처리 무조건 좋은가?

### DISCUSSION THREAD

**Discussion by JoisFe** [cowaine/spring-boot-for-msa#44] [Chapter 05. REST APIs with Spring MVC] [answered]
프로그래밍 공부를 하면서 자주 들은 말이
`예외처리 회피는 정말 필요한 경우에 사용하는 것이 좋다`
그 이유로는 예외처리를 회피하면 다른 곳에서 예외 처리를 하면서 코드 복잡성이 생기거나, 정확한 예외가 일어난 지점을 파악하는데 어려운 (디버깅) 상황이 생기기 때문으로 알 고 있습니다.
하지만 `@ControllerAdvice`, `@ExceptionHandler` 를사용하면 예외가 발생하는 곳에서 예외처리를 하지 않고 전역적으로 예외를 처리하는 것을 볼 수 있습니다.
예외 관리를 한 곳에서 하니 매우 편하여 코드 복잡성은 오히려 줄어드는 것 같고, 예외가 일어난 지점을 찾는 것은 throw를 한 곳을 IDE를 디버깅 모드를 이용해 이전 단계를 물어가며 찾아가는 것은 그리 어렵지 않다고 느껴지기 집니다.
괜히 `예외처리 회피는 정말 필요한 경우에 사용하는 것이 좋다` 이러한 문구가 생긴게 아닌 것 같은데
`@ControllerAdvice`를 사용하지 않고 예외가 발생한 지점에서 바로 잡는 것이 좋은 예시가 있을까요? 또한 그러한 경우를 판단하기 쉽게 일반화? 할 수 있을까요?

**Comment by corock** [comment, marked as answer]
이게 적절한 예시인지는 모르겠으나... 특정 아이템의 재고를 차감하는 로직을 예로 들어보겠습니다.
오히려 예시를 보고 반대로 질문을 드리는 게 좋을 것 같네요~!

> A 시스템의 재고 차감은 가 차감 방식으로 되어 있고, 공유 자원인 아이템에 대해 고객이 락(_lock_)을 획득하고 아이템 가용 수량을 확보하여 구매한다.
> 만약 트래픽이 몰리는 시점에 후순위로 진입한 고객은 락 획득에 실패하게 되고, 해당 고객은 일정 기간을 두고 2회 재시도한다.
> 재수 없게도 2회 모두 락 획득에 실패할 경우 검사 예외(_checkued exception_)인 `InterruptedException` 이 발생한다.

이런 상황에는 어떻게 예외 처리를 하면 좋을까요?
`@ControllerAdvice` 이나 `@ExceptionHandler` 애너테이션을 사용하면 적절한 상황인지 궁금합니다.
참고로 해당 케이스는 현재 시스템에서 재고 차감하는 로직 한 군데에서 발생하고 있다고 가정합니다.
